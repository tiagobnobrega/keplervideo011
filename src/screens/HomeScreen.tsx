import {
  ChannelServerComponent,
  IChangeChannelResponse,
} from '@amzn/kepler-channel';
import {
  ContentPersonalizationServer,
  CustomerListType,
} from '@amzn/kepler-content-personalization';
import {
  ContentLauncherServerComponent,
  ContentLauncherStatusType,
  IContentLauncherHandler,
  IContentSearch,
  ILaunchContentOptionalFields,
  ILauncherResponse,
} from '@amzn/kepler-media-content-launcher';
import { getMockedCurrentTitleDataForChannel } from '../livetv/mock/MockSource';

import { SubscriptionEntitlementServer } from '@amzn/kepler-subscription-entitlement';
import { useFocusEffect } from '@amzn/react-navigation__core';
import { isEqual } from 'lodash';
import React, { useCallback, useEffect, useState } from 'react';
import { StyleSheet, View } from 'react-native';
import { EventRegister } from 'react-native-event-listeners';
import { useDispatch, useSelector } from 'react-redux';
import {
  AccountLoginWrapperInstance,
  onStartService as onStartAccountLoginService,
  onStopService as onStopAccountLoginService,
} from '../AccountLoginWrapper';
import ContentPreview from '../components/ContentPreview';
import MovieGrid from '../components/MovieGrid';
import { AppStackScreenProps, Screens } from '../components/navigation/types';
import { OptionType } from '../components/RadioPicker';
import AutoRotator from '../components/rotator/AutoRotator';
import { MovieRotatorGrid } from '../components/touchOptimized/MovieRotatorGrid';
import {
  isAccountLoginEnabled,
  isContentPersonalizationEnabled,
  isDpadControllerSupported,
  isSubscriptionEntitlementEnabled,
} from '../config/AppConfig';
import { LoginStatus } from '../constants';
import { getClassics } from '../data/local/classics';
import { getLatestHits } from '../data/local/latestHits';
import { getNewHits } from '../data/local/newHits';
import { getRecommendations } from '../data/local/recommendations';
import { getRotator } from '../data/local/rotator';
import { getTestAssets } from '../data/local/testAssests';
import { getTrends } from '../data/local/trends';
import { tileData } from '../data/tileData';
import channelTunerHandler from '../livetv/channelTunerHandler';
import { setCurrentFocus } from '../store/focus/focusSlice';
import {
  setCountryCode,
  settingsSelectors,
} from '../store/settings/SettingsSlice';
import { COLORS } from '../styles/Colors';
import { MovieGridData } from '../types/MovieGridData';
import { TitleData } from '../types/TitleData';
import { recordCustomEvent } from '../utils/helperNewRelic';
import { scaleUxToDp } from '../utils/pixelUtils';
import { getSelectedLocale } from '../utils/translationHelper';

const AutoRotatorData = getRotator();

const data: MovieGridData[] = [
  {
    heading: 'Latest Hits',
    testID: 'costa_rica_top_rated',
    data: getLatestHits,
  },
  {
    heading: 'Classics',
    testID: 'costa_rica_attractions',
    data: getClassics,
  },
  {
    heading: 'Recommendation',
    testID: 'costa_rica_underwater',
    data: getRecommendations,
  },
  {
    heading: 'Test Assets',
    testID: 'test_assets',
    data: getTestAssets,
  },
  {
    heading: 'New hits',
    testID: 'new_hits',
    data: getNewHits,
  },
  {
    heading: 'Trends',
    testID: 'trends',
    data: getTrends,
  },
];

interface HomeProps {
  navigation: any;
}

const HomeScreen = ({
  navigation,
}: AppStackScreenProps<Screens.HOME_SCREEN>) => {
  const dispatch = useDispatch();
  const countryCode = useSelector(settingsSelectors.countryCode);
  const [selectedTitle, setSelectedTitle] = useState<TitleData | undefined>(
    undefined,
  );
  const firstTileRef = React.useRef<View>(null);
  const [destination, setDestination] = useState<any>([]);
  const destinations = destination?.current ? [destination?.current] : [];
  let listener: string | boolean;
  const NAVIGATION_DELAY = 700;

  if (!countryCode) {
    const selectedLocale = getSelectedLocale();
    if (selectedLocale) {
      dispatch(setCountryCode(selectedLocale as OptionType));
    }
  }
  const callCustomerListRefresh = () => {
    if (!isContentPersonalizationEnabled()) {
      return;
    }
    console.log('k_content_per: Calling Report Refreshed Customer List');
    ContentPersonalizationServer.reportRefreshedCustomerList(
      CustomerListType.WATCHLIST,
    );
  };

  useFocusEffect(
    useCallback(() => {
      dispatch(
        setCurrentFocus({
          currentFocusedScreen: Screens.HOME_SCREEN,
        }),
      );
      return () => {
        dispatch(
          setCurrentFocus({
            currentFocusedScreen: '',
          }),
        );
      };
      // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []),
  );

  const callContentEntitlementsRefresh = useCallback(() => {
    if (!isContentPersonalizationEnabled()) {
      return;
    }
    console.log('k_content_per: Calling Report Refreshed Content Entitlements');
    ContentPersonalizationServer.reportRefreshedContentEntitlements();
  }, []);

  const callPlaybackEventsRefresh = useCallback(() => {
    if (!isContentPersonalizationEnabled()) {
      return;
    }
    console.log('k_content_per: Calling Report Refreshed PlaybackEvents');
    ContentPersonalizationServer.reportRefreshedPlaybackEvents();
  }, []);

  const callSubscriptionEntitlementsRefresh = useCallback(() => {
    if (!isSubscriptionEntitlementEnabled()) {
      return;
    }
    console.log(
      'k_sub_ent: Calling Report Refreshed Subscription Entitlements',
    );
    SubscriptionEntitlementServer.reportRefreshedSubscriptions();
  }, []);

  const navigateToPlayer = useCallback(
    (changeChannelResponse?: IChangeChannelResponse) => {
      navigation.navigate(Screens.HOME_SCREEN);
      setTimeout(async () => {
        // Try to match the changeChannelResponseData with a specific EPG program.
        const mockedData = await getMockedCurrentTitleDataForChannel(
          changeChannelResponse?.data ?? '',
        );
        const videoData = { ...tileData, ...mockedData };
        const params = {
          data: videoData,
          sendDataOnBack: () => {},
        };
        try {
          navigation.navigate(Screens.PLAYER_SCREEN, params);
          recordCustomEvent('Navigated: ', {
            name: Screens.PLAYER_SCREEN,
            params: videoData,
          });
        } catch {
          navigation.goBack();
        }
        // Simulating load time.
      }, NAVIGATION_DELAY);
    },
    [navigation],
  );

  const onTileFocus = useCallback(
    (title?: TitleData) => {
      setSelectedTitle(title);
      setDestination(null);
    },
    [setSelectedTitle],
  );

  const setNextFocusDestination = () => {
    setDestination(firstTileRef);
  };

  const setFocusDestinationFromRotator = () => {
    firstTileRef?.current?.focus();
    setDestination(firstTileRef);
  };

  useEffect(() => {
    ChannelServerComponent.channelServer.handler = channelTunerHandler;
    callCustomerListRefresh();
    callContentEntitlementsRefresh();
    callPlaybackEventsRefresh();
    callSubscriptionEntitlementsRefresh();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    const factory = new ContentLauncherServerComponent();
    const contentLauncherHandler: IContentLauncherHandler = {
      async handleLaunchContent(
        contentSearch: IContentSearch,
        autoPlay: boolean,
        _optionalFields: ILaunchContentOptionalFields,
      ): Promise<ILauncherResponse> {
        console.log(
          'HeadlessLaunchContentHandler handleLaunchContent invoked.',
        );

        // iterator to print
        let searchParameters = contentSearch.getParameterList();
        if (searchParameters.length > 0) {
          console.log(
            `Content Launcher: Search param List Length: ${searchParameters.length}`,
          );

          let searchString = '';
          for (var j = 0; j < searchParameters.length; j++) {
            let additionalInfoList = searchParameters[j].getExternalIdList();
            console.log(
              `Content Launcher: additionalInfoList.length : ${additionalInfoList.length}`,
            );
            for (var i = 0; i < additionalInfoList.length; i++) {
              searchString += '\n';
              searchString += additionalInfoList[i].getName();
              searchString += ' : ';
              searchString += additionalInfoList[i].getValue();
              console.log(
                `Content Launcher: Search Str in additionalInfoList  ${i} : ${searchString}`,
              );
            }
            searchString += '\n\n';
            console.log(`Content Launcher: Final Search str ${searchString}`);
          }

          console.log('Content Launcher: Going to tell launch type');
          if (autoPlay) {
            console.log(`Content Launcher: Quickplay ${searchString}`);
          } else {
            console.log(`Content Launcher: In-App search ${searchString}`);
          }
        } else {
          console.log('Content Launcher: Error fetching search string');
        }
        //iterator to print

        console.log(
          'HeadlessLaunchContentHandler handleLaunchContent invoked.',
        );

        const launcherResponse = factory
          .makeLauncherResponseBuilder()
          .contentLauncherStatus(ContentLauncherStatusType.SUCCESS)
          .build();

        if (
          launcherResponse.getContentLauncherStatus() ===
          ContentLauncherStatusType.SUCCESS
        ) {
          navigateToPlayer();
        }
        return Promise.resolve(launcherResponse);
      },
    };
    const contentLauncherServer = factory.getOrMakeServer();
    contentLauncherServer.setHandler(contentLauncherHandler);
    // eslint-disable-next-line react-hooks/exhaustive-deps
    listener = EventRegister.addEventListener(
      'LiveChannelEvent',
      navigateToPlayer,
    );

    return () => {
      EventRegister.removeEventListener(listener as string);
    };
  }, []);

  useEffect(() => {
    if (!isAccountLoginEnabled()) {
      return;
    }

    const startAccountLoginInstance = async () => {
      await onStartAccountLoginService();
      AccountLoginWrapperInstance.updateStatus(LoginStatus.SIGNED_IN);
    };

    startAccountLoginInstance();

    return () => {
      onStopAccountLoginService();
    };
  }, []);

  if (!isDpadControllerSupported()) {
    // Render touch optimized Home Screen that scrolls banner and video tiles
    // as one unit.
    return (
      <View style={styles.movieGridContainer}>
        <MovieRotatorGrid rotatorData={AutoRotatorData} movieGridData={data} />
      </View>
    );
  }

  // Render DPad optimized Home Screen that utilizes the banner area to show
  // content preview for the in-focus video tile.
  return (
    <View style={styles.container}>
      <View style={styles.contentPreviewAndRotatorContainer}>
        {selectedTitle ? (
          <ContentPreview
            tile={selectedTitle}
            onFocus={setNextFocusDestination}
          />
        ) : (
          <AutoRotator
            data={AutoRotatorData}
            onFocus={setFocusDestinationFromRotator}
          />
        )}
      </View>
      <View style={styles.movieGridContainer}>
        <MovieGrid
          data={data}
          initialColumnsToRender={6}
          initialRowsToRender={3}
          onTileFocus={onTileFocus}
          ref={firstTileRef}
          testID={'movieGrid'}
          destinations={destinations}
        />
      </View>
    </View>
  );
};

const homeScreenDataUnchanged = (
  prevProps: HomeProps,
  nextProps: HomeProps,
) => {
  return isEqual(prevProps, nextProps);
};

export default React.memo(HomeScreen, homeScreenDataUnchanged);

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: COLORS.BLACK,
    paddingLeft: scaleUxToDp(29),
  },
  contentPreviewAndRotatorContainer: {
    flex: 0.8,
    width: '100%',
    height: '100%',
  },
  movieGridContainer: {
    flex: 1,
    width: '100%',
    height: '100%',
    backgroundColor: COLORS.BLACK,
    marginBottom: 20,
  },
});
