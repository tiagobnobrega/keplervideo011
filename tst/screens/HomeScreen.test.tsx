import { ContentLauncherStatusType } from '@amzn/kepler-media-content-launcher';
import { LinearGradientProps } from '@amzn/react-linear-gradient';
import { render } from '@testing-library/react-native';
import React from 'react';
import { EventRegister } from 'react-native-event-listeners';
import { useDispatch } from 'react-redux';
import {
  AppStackScreenProps,
  Screens,
} from '../../src/components/navigation/types';
import HomeScreen from '../../src/screens/HomeScreen';
import {
  getSelectedLocale,
  localeOptions,
} from '../../src/utils/translationHelper';

jest.mock('@amzn/react-linear-gradient', () => ({
  __esModule: true,

  default: ({ children }: LinearGradientProps) => <>{children}</>,
}));

jest.mock('@amzn/kepler-media-content-launcher', () => ({
  ContentLauncherServerComponent: jest
    .fn()
    .mockImplementation(() => mockContentLauncherServerComponent),
  ContentLauncherStatusType: {
    SUCCESS: 'SUCCESS',
  },
}));

jest.mock('lodash', () => ({
  isEqual: jest.fn().mockReturnValue(true),
}));

jest.mock('../../src/data/videos', () => ({
  DEFAULT_FILE_TYPE: 'video/mp4',
}));

const mockContentLauncherServerComponent = {
  makeLauncherResponseBuilder: jest.fn().mockReturnThis(),
  contentLauncherStatus: jest.fn().mockReturnThis(),
  build: jest.fn().mockReturnValue({
    getContentLauncherStatus: jest
      .fn()
      .mockReturnValue(ContentLauncherStatusType.SUCCESS),
  }),
  getOrMakeServer: jest.fn().mockReturnThis(),
  setHandler: jest.fn(),
};

const mockNavigation: any = {
  navigate: jest.fn(),
  goBack: jest.fn(),
};
const mockRoute: any = {};

const props: AppStackScreenProps<Screens.HOME_SCREEN> = {
  navigation: mockNavigation,
  route: mockRoute,
};

describe('HomeScreen', () => {
  beforeAll(() => {
    jest.useFakeTimers();
  });

  let mockDispatch: any;
  beforeEach(() => {
    mockDispatch = jest.fn();
    (useDispatch as unknown as jest.Mock).mockReturnValue(mockDispatch);
  });

  afterAll(() => {
    jest.useRealTimers();
  });

  it('renders correctly and matches snapshot', () => {
    const { toJSON } = render(<HomeScreen {...props} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('registers event listener and handles event properly', () => {
    const { unmount } = render(<HomeScreen {...props} />);
    expect(EventRegister.addEventListener).toHaveBeenCalledWith(
      'LiveChannelEvent',
      expect.any(Function),
    );
    const eventHandler = (
      EventRegister.addEventListener as unknown as jest.Mock
    ).mock.calls[0][1];
    eventHandler();
    unmount();
    expect(EventRegister.removeEventListener).toHaveBeenCalledWith(
      'mockListenerId',
    );
  });
});

describe('HomeScreen with React.memo', () => {
  it('renders only when props change', async () => {
    const isEqualSpy = jest.spyOn(require('lodash'), 'isEqual');
    const { rerender } = render(<HomeScreen {...props} />);
    rerender(<HomeScreen {...props} />);
    expect(isEqualSpy).toHaveBeenCalledTimes(1);
    const updatedProps = { ...props, someNewProp: 'new value' };
    rerender(<HomeScreen {...updatedProps} />);
    expect(isEqualSpy).toHaveBeenCalledWith(props, updatedProps);
    rerender(<HomeScreen {...updatedProps} />);
    expect(isEqualSpy).toHaveBeenCalledTimes(3);
    isEqualSpy.mockRestore();
  });
});

describe('HomeScreen with touch optimized UX', () => {
  beforeEach(() => {
    jest
      .spyOn(require('../../src/config/AppConfig'), 'isDpadControllerSupported')
      .mockReturnValue(false);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly and matches snapshot', () => {
    const { toJSON } = render(<HomeScreen {...props} />);
    expect(toJSON()).toMatchSnapshot();
  });
});

describe('HomeScreen with touch optimized UX And Disabled Additional features', () => {
  const ContentPersonalizationServer =
    require('@amzn/kepler-content-personalization').ContentPersonalizationServer;
  const AccountLoginWrapperInstance =
    require('../../src/AccountLoginWrapper').AccountLoginWrapperInstance;
  const onStartService =
    require('../../src/AccountLoginWrapper').onStartService;
  const onStopService = require('../../src/AccountLoginWrapper').onStopService;

  beforeEach(() => {
    jest.mock('@amzn/kepler-content-personalization', () => ({
      __esModule: true,

      ContentPersonalizationServer: jest.fn(),
    }));

    jest.mock('../../src/AccountLoginWrapper', () => ({
      __esModule: true,

      AccountLoginWrapperInstance: jest.fn(),
      onStartService: jest.fn(),
      onStopService: jest.fn(),
    }));

    jest
      .spyOn(require('../../src/config/AppConfig'), 'isDpadControllerSupported')
      .mockReturnValue(false);
    jest
      .spyOn(
        require('../../src/config/AppConfig'),
        'isContentPersonalizationEnabled',
      )
      .mockReturnValue(false);
    jest
      .spyOn(
        require('../../src/config/AppConfig'),
        'isSubscriptionEntitlementEnabled',
      )
      .mockReturnValue(false);
    jest
      .spyOn(require('../../src/config/AppConfig'), 'isAccountLoginEnabled')
      .mockReturnValue(false);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('does not exercise content personalization features', () => {
    render(<HomeScreen {...props} />);
    expect(
      ContentPersonalizationServer.reportRefreshedCustomerList,
    ).toHaveBeenCalledTimes(0);
    expect(
      ContentPersonalizationServer.reportRefreshedContentEntitlements,
    ).toHaveBeenCalledTimes(0);
    expect(
      ContentPersonalizationServer.reportRefreshedPlaybackEvents,
    ).toHaveBeenCalledTimes(0);
  });

  it('does not exercise account login features', () => {
    render(<HomeScreen {...props} />);
    expect(AccountLoginWrapperInstance.updateStatus).toHaveBeenCalledTimes(0);
    expect(onStartService).toHaveBeenCalledTimes(0);
    expect(onStopService).toHaveBeenCalledTimes(0);
  });
});

describe('getSelectedLocale', () => {
  it('should return a valid OptionType object', () => {
    const result = getSelectedLocale();

    expect(result).toBeDefined();
    expect(result).toHaveProperty('code');
    expect(result).toHaveProperty('label');
    expect(result).toHaveProperty('value');

    expect(result).toMatchObject({
      code: expect.any(String),
      label: expect.any(String),
      value: expect.any(String),
    });
  });

  it('should return the default OptionType if currentCountry is undefined', () => {
    jest.mock('../../src/utils/translationHelper', () => ({
      getSelectedLocale: jest.fn().mockReturnValue(undefined),
    }));

    const result = getSelectedLocale();

    expect(result).toEqual(localeOptions[0]);
  });
});
