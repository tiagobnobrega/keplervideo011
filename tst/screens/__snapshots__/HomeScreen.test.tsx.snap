// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen renders correctly and matches snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#000000",
      "flex": 1,
      "paddingLeft": 29,
    }
  }
>
  <View
    style={
      Object {
        "flex": 0.8,
        "height": "100%",
        "width": "100%",
      }
    }
  >
    <AutoRotator
      data={
        Array [
          Object {
            "categories": Array [
              "Hits",
            ],
            "channelID": "13454",
            "description": "Lorem ipsum dolor sit amet",
            "duration": "124",
            "id": "169328",
            "posterUrl": undefined,
            "rating": "4",
            "releaseDate": "09/20",
            "thumbnailUrl": undefined,
            "title": "Scrimshun",
            "videoUrl": "https://edge-vod-media.cdn01.net/encoded/0000169/0169319/video_1880k/A57CCB0FI.mp4?source=firetv&channelID=13455",
          },
          Object {
            "categories": Array [
              "Hits",
            ],
            "channelID": "13454",
            "description": "Lorem ipsum dolor sit amet",
            "duration": "48",
            "id": "169329",
            "posterUrl": undefined,
            "rating": "4",
            "releaseDate": "09/20",
            "thumbnailUrl": undefined,
            "title": "Dreamcatcher",
            "videoUrl": "https://edge-vod-media.cdn01.net/encoded/0000169/0169313/video_1880k/T7J66Z106.mp4?source=firetv&channelID=13454",
          },
          Object {
            "categories": Array [
              "Hits",
            ],
            "channelID": "13453",
            "description": "Lorem ipsum dolor sit amet",
            "duration": "185",
            "id": "169330",
            "posterUrl": undefined,
            "rating": "4",
            "releaseDate": "09/20",
            "thumbnailUrl": undefined,
            "title": "Twinsouls",
            "videoUrl": "https://edge-vod-media.cdn01.net/encoded/0000169/0169313/video_1880k/T7J66Z106.mp4?source=firetv&channelID=13454",
          },
          Object {
            "categories": Array [
              "Hits",
            ],
            "channelID": "13455",
            "description": "Lorem ipsum dolor sit amet",
            "duration": "156",
            "id": "169331",
            "posterUrl": undefined,
            "rating": "4",
            "releaseDate": "09/20",
            "thumbnailUrl": undefined,
            "title": "Midnight Masquerade",
            "videoUrl": "https://edge-vod-media.cdn01.net/encoded/0000169/0169319/video_1880k/A57CCB0FI.mp4?source=firetv&channelID=13455",
          },
        ]
      }
      onFocus={[Function]}
    />
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#000000",
        "flex": 1,
        "height": "100%",
        "marginBottom": 20,
        "width": "100%",
      }
    }
  >
    <MovieGrid
      data={
        Array [
          Object {
            "data": [Function],
            "heading": "Latest Hits",
            "testID": "costa_rica_top_rated",
          },
          Object {
            "data": [Function],
            "heading": "Classics",
            "testID": "costa_rica_attractions",
          },
          Object {
            "data": [Function],
            "heading": "Recommendation",
            "testID": "costa_rica_underwater",
          },
          Object {
            "data": [Function],
            "heading": "Test Assets",
            "testID": "test_assets",
          },
          Object {
            "data": [Function],
            "heading": "New hits",
            "testID": "new_hits",
          },
          Object {
            "data": [Function],
            "heading": "Trends",
            "testID": "trends",
          },
        ]
      }
      destinations={Array []}
      initialColumnsToRender={6}
      initialRowsToRender={3}
      onTileFocus={[Function]}
      testID="movieGrid"
    />
  </View>
</View>
`;

exports[`HomeScreen with touch optimized UX renders correctly and matches snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#000000",
      "flex": 1,
      "height": "100%",
      "marginBottom": 20,
      "width": "100%",
    }
  }
>
  <RCTScrollView
    canChangeSize={false}
    contentHeight={0}
    contentWidth={0}
    dataProvider={
      DataProvider {
        "_data": Array [
          Object {
            "data": [Function],
            "heading": "Costa Rica Attractions",
            "testID": "grid-rotator",
            "type": "rotator",
          },
          Object {
            "data": [Function],
            "heading": "Latest Hits",
            "testID": "costa_rica_top_rated",
            "type": "carousel",
          },
          Object {
            "data": [Function],
            "heading": "Classics",
            "testID": "costa_rica_attractions",
            "type": "carousel",
          },
          Object {
            "data": [Function],
            "heading": "Recommendation",
            "testID": "costa_rica_underwater",
            "type": "carousel",
          },
          Object {
            "data": [Function],
            "heading": "Test Assets",
            "testID": "test_assets",
            "type": "carousel",
          },
          Object {
            "data": [Function],
            "heading": "New hits",
            "testID": "new_hits",
            "type": "carousel",
          },
          Object {
            "data": [Function],
            "heading": "Trends",
            "testID": "trends",
            "type": "carousel",
          },
        ],
        "_firstIndexToProcess": 0,
        "_hasStableIds": false,
        "_requiresDataChangeHandling": false,
        "_size": 7,
        "getStableId": [Function],
        "rowHasChanged": [Function],
      }
    }
    disableRecycling={false}
    externalScrollView={[Function]}
    horizontal={false}
    initialNumToRender={1}
    initialOffset={0}
    initialRenderIndex={0}
    isHorizontal={false}
    layoutProvider={
      LayoutProvider {
        "_getLayoutTypeForIndex": [Function],
        "_setLayoutForType": [Function],
        "_tempDim": Object {
          "height": 0,
          "width": 0,
        },
        "shouldRefreshWithAnchoring": true,
      }
    }
    onEndReachedThreshold={0}
    onEndReachedThresholdRelative={0}
    onLayout={[Function]}
    onScroll={[Function]}
    onSizeChanged={[Function]}
    removeClippedSubviews={false}
    renderAheadOffset={250}
    rowRenderer={[Function]}
    scrollEventThrottle={16}
    scrollThrottle={16}
    testID="movie-rotator-grid"
    useWindowScroll={true}
  >
    <View>
      <View
        style={
          Object {
            "flexDirection": "column",
          }
        }
      >
        <View
          horizontal={false}
          renderAheadOffset={250}
          scrollOffset={0}
          style={
            Object {
              "height": 0,
              "width": 0,
            }
          }
          windowSize={250}
        />
      </View>
    </View>
  </RCTScrollView>
</View>
`;
